### PR TITLE
[New Feature] Support LARS and code refactor DistriOptimizer for advance parameter operations 

### PR DESCRIPTION
**Pull request by dding3**
## What changes were proposed in this pull request?
1. support LARS
2. code refactor for various parameter operations in DistriOptimizer
3. add ut for gradient clipping
4. update parameter type from float to double in Tensor.clamp to be consistent with other tensor apis

## How was this patch tested?
manual test and ut

### CODE REVIEW

**Comment by yiheng** [comment]
I think the DistriOptimizer file is too long to maintain now...

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala [hunk 1]
```diff
@@ -825,5 +825,5 @@ trait TensorMath[T] {
 
   def sumSquare(): T
 
-  def clamp(min: Float, max: Float): Tensor[T]
+  def clamp(min: Double, max: Double): Tensor[T]
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala]: why change this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 2]
```diff
@@ -134,7 +138,8 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int, size: Int) ex
    * @param parameter A tensor representing the initial underlying weights of this
    *                  `AllReduceParameter`
    */
-  def init(parameter: Tensor[T])(implicit ev: TensorNumeric[T]): Unit = {
+  def init(parameter: Tensor[T], sizes: Array[Int])(implicit ev: TensorNumeric[T]):
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: What's the sizes for? can you update the java doc?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 3]
```diff
@@ -134,7 +138,8 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int, size: Int) ex
...
+    (Int, Int, Int) = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: what's the result mean?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala [hunk 4]
```diff
@@ -304,6 +311,20 @@ class AllReduceParameter[T: ClassTag](id: Long, partitionNum: Int, size: Int) ex
       .getOrElse(throw new IllegalStateException("Please initialize AllReduceParameter first!"))
     weights.copy(weightPartition)
   }
+
+  def squarePerLayer(lookupDic: Array[mutable.HashMap[Int, (Int, Int)]]):
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: It's a public API? do you need add java doc for this method???

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameter.scala]: Why AllReduceParameter need to know the ```Layer```???

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 5]
```diff
@@ -168,13 +257,17 @@ object DistriOptimizer {
...
-    val normValueClip = clippingParams.normValueClip
-
+    val advanceOperationList = new ArrayBuffer[String]()
+    if (clippingParams.enableL2NormClipping) {
+      advanceOperationList += "norm2Clip"
+    }
+    if (clippingParams.enableConstantClipping) {
+      advanceOperationList += "constantClip"
+    }
+    if (optimMethod.isInstanceOf[LarsSGD[T]]) {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: It's not good to do something special for a specific OptimMethod in Optimizer.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 6]
```diff
@@ -66,6 +67,93 @@ object DistriOptimizer {
...
+    metrics: Metrics,
+    lookupDicBroadcast: Broadcast[Array[mutable.HashMap[Int, (Int, Int)]]])
+    (implicit ev: TensorNumeric[T]):
+    (Boolean, Double, Broadcast[Array[(Int, (Double, Double))]]) = {
+    var aggregatedG: Boolean = false
+    // array element is a tuple, first element is layer id, value is (wNorm2, gNorm2)
+    var gwNorm2ListBroadcast: Broadcast[Array[(Int, (Double, Double))]] = null
+    var l2Norm = 0.0
+    operations.foreach { oper =>
+      if (oper.equals("norm2Clip")) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: so if we add another string, we also need to change the ```if else``` blocks? Why not wrap them with class...

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 7]
```diff
@@ -635,6 +698,67 @@ object DistriOptimizer {
         }
       }
       Engine.setNodeAndCore(nExecutor, executorCores)
+      val sizeValues = sizesBroadcast.value
+      val weights = modelBroadcast.value(false).getParameters()._1
+      Iterator.single(parameters.init(weights, sizeValues))
+    }.collect()
+
+    var lookupDic: Array[mutable.HashMap[Int, (Int, Int)]] = null
+    if (optimMethod.isInstanceOf[LarsSGD[T]]) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: can we move this part to the collectGlobalData method?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 8]
```diff
@@ -66,6 +67,93 @@ object DistriOptimizer {
...
+    (implicit ev: TensorNumeric[T]):
+    (Boolean, Double, Broadcast[Array[(Int, (Double, Double))]]) = {
+    var aggregatedG: Boolean = false
+    // array element is a tuple, first element is layer id, value is (wNorm2, gNorm2)
+    var gwNorm2ListBroadcast: Broadcast[Array[(Int, (Double, Double))]] = null
+    var l2Norm = 0.0
+    operations.foreach { oper =>
+      if (oper.equals("norm2Clip")) {
+        val sumSquare = models.mapPartitions(modelIter => {
+          val getG = System.nanoTime()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Can we move these logic into optimMethod? We can expand optimMethod API

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 9]
```diff
@@ -664,10 +786,9 @@ object DistriOptimizer {
     logger.info("Cache thread models...")
     models.count()
     logger.info("Cache thread models... done")
-    models
+    (models, sc.broadcast(lookupDic))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: This is...

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 10]
```diff
@@ -309,68 +402,37 @@ object DistriOptimizer {
...
-          if (!normClippingEnable) {
+          if (!aggregateG) {
             val getG = System.nanoTime()
-            parameters.aggregateGradientPartition()
+            parameters.aggregateGradientPartition(numFinishedModelUpdatesT)
             driverMetrics.add("aggregrateGradientParition average executor",
               System.nanoTime() - getG)
           }
-          parameters.gradientPartition.div(scale)
+            advanceOperations(modelCache, parameters, clippingParams, l2Norm,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: what's the advanceOperation mean? What should we expect from it? What should we put into it.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 11]
```diff
@@ -105,7 +193,8 @@ object DistriOptimizer {
     trainSummary: Option[TrainSummary],
     validationSummary: Option[ValidationSummary],
     isOverWrite: Boolean,
-    clippingParams: GradientClippingParams
+    clippingParams: GradientClippingParams,
+    lookupDicBroadcast: Broadcast[Array[mutable.HashMap[Int, (Int, Int)]]]
```
[the remark this answers is not in the document]
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Updated

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala (hunk 1 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala]: Updated in the pull request

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 10 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Update the comments for this api

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 9 above)
Comment (dding3) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Update in the comments

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala [hunk 12]
```diff
@@ -102,6 +103,13 @@ trait OptimMethod[@specialized(Float, Double) T] extends Serializable {
    */
   def loadFromTable(config: Table): this.type
 
+  /**
+   * get parameter processor
+   *
+   * @return
+   */
+  def getParameterProcessor(): ParameterProcessor = null
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala]: Option[...]?

**Comment by yiheng** [comment]
Add some tests for LarsSGD and ParameterOperations?

**Comment by yiheng** [comment]
others looks good to me

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 13]
```diff
@@ -371,26 +376,28 @@ abstract class Optimizer[T: ClassTag, D](
...
-    gradientClippingParams.enableL2NormClipping = true
-    gradientClippingParams.normValueClip = clipNorm
+    parameterProcessors.append(new L2NormClippingProcessor(l2NormThreshold))
     this
   }
+
+  /**
+   * a list of ParameterProcessor, orders matter
+   */
+  protected var parameterProcessors = ArrayBuffer[ParameterProcessor]()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Why not use a LinkedList  instead of ArrayBuffer？

**Comment by yiheng** [comment]
Update in another PR https://github.com/intel-analytics/BigDL/pull/2556

**Pull request closed without merging**